src/xopen/__init__.py

```diff
@@ -356,6 +356,46 @@ def _open_gz(filename, mode, compresslevel, threads):
         return buffered_writer(gzip.open(filename, mode, compresslevel=compresslevel))
 
 
+def _detect_format_from_content(filename, mode, compresslevel, threads):
+    """
+    Attempts to detect file format from the content.
+    If the file mode isn't a 'read' mode or the file isn't a 'regular' file, function
+    returns None. Otherwise, we return the open function of the format we've detected.
+    """
+    try:
+        if mode.startswith("r") and stat.S_ISREG(os.stat(filename).st_mode):
+            with open(filename, "rb") as fh:
+                bs = fh.read(6)
+                if not _PY3:
+                    bs = bytearray(bs)
+            if bs[:2] == b'\x1f\x8b':
+                # https://tools.ietf.org/html/rfc1952#page-6
+                return _open_gz(filename, mode, compresslevel, threads)
+            elif bs[:3] == b'\x42\x5a\x68':
+                # https://en.wikipedia.org/wiki/List_of_file_signatures
+                return _open_bz2(filename, mode)
+            elif bs[:6] == b'\xfd\x37\x7a\x58\x5a\x00' and _PY3:
+                # lzma module is not available for python 2.7
+                # https://tukaani.org/xz/xz-file-format.txt
+                return _open_xz(filename, mode)
+    except OSError as e:
+        # users expect IOError not OSError
+        raise IOError(e)
```

Comment:
As I wrote above, you should just return `None`, here but I wanted to mention anyway that `IOError` is an alias of `OSError` [starting from Python 3.3](https://docs.python.org/3/library/exceptions.html#OSError).